I18n

patkkka opened the pull request:
I've moved dao, models and service to backend module, however now I don't know how to run this application correctly. I've created GymBackendApplication in backend module and now app is working only when:
1) GymBackendApplication is started and it immediately finishes beacuse there is no controller
2) then GymApplication is started

JSON example in request body to i18n controller is:
{
	"languageCode" : "pl_PL",
	"messagesKeys" : ["MK1", "MK2"]
}

patkkka committed:
Add i18n rest controller, service, repositories for locale and user messages

patkkka committed:
Move model, dao & serivce packages to backend

patkkka committed:
Change controller method to POST and to retrieving request body with language code and messages keys, change service method to find messages based on both language code and message key

patkkka committed:
Implement test for controller method

patkkka committed:
Implement equals and hashCode tests for entities

patkkka committed:
Add test for i18n service method

arturAR commented on backend/src/main/java/com/java/academy/gym/model/UserMessage.java:
```diff
@@ -0,0 +1,73 @@
+
+@Entity
+@Table(name = "userMessages")
+public class UserMessage extends BaseEntity{
```
space after class name

arturAR commented on backend/src/main/java/com/java/academy/gym/model/UserMessage.java:
```diff
@@ -0,0 +1,73 @@
+    private Locale locale;
+
+    @NotNull
+    @Size(max = 100)
```
Message can be longer than 100. Minimum is 255(in my opinion it could be ~500)

patkkka replied:
This is only message key, for message text I didn't set any max size.

arturAR replied:
Ok, but still, key can be longer than 100.

patkkka replied:
Okk

arturAR commented on backend/src/main/java/com/java/academy/gym/repository/LocaleDao.java:
```diff
@@ -0,0 +1,11 @@
+public interface LocaleDao extends JpaRepository<Locale, Long> {
+
+    Locale findByLanguageCode(String languageCode);
+}
```
Empty line here.

arturAR commented on backend/src/test/java/com/java/academy/gym/model/UserMessageTest.java:
```diff
@@ -0,0 +1,38 @@
+    @Test
+    public void shouldReturnThatUserMessagesAreEqual() {
+        Locale locale1 = new Locale("pl_PL", "polish", "Poland");
+        Locale locale2 = new Locale("pl_PL", "polish", "Poland");
```
In integration tests this test could not pass. Why? because "@Column(unique = true)"

mprtcz replied:
Then it is a good thing that it is only a unit test :P

patkkka replied:
@arturAR I don't understand, this is only unit test as @mprtcz said. Do you want to exclude languageCode from equals method or what do yo mean?

volodymyrd replied:
yes it's just a unit test, it doesn't care about database constraints, but for this purpose more valuable is a integration test

arturAR commented on backend/src/test/java/com/java/academy/gym/model/UserMessageTest.java:
```diff
@@ -0,0 +1,38 @@
+
+    @Test
+    public void shouldReturnThatUserMessagesAreDifferent() {
+        UserMessage userMessage1 = new UserMessage(null, "MK1", "Message1");
```
In integration tests this test could not pass. Why? because "nullable = false"

arturAR commented on backend/src/test/java/com/java/academy/gym/service/impl/InternationalizationServiceImplTest.java:
```diff
@@ -0,0 +1,51 @@
+        Map<String, String> userMessages = internationalizationService
+                .findUserMessagesByKeysAndLanguageCode(messagesKeys, languageCode);
+
+        Map<String, String> expectedUserMessages = new HashMap<>();
```
assertTrue(contains(message)) - better than assertEquals in this case

arturAR commented on frontend/src/main/java/com/java/academy/gym/controller/InternationalizationController.java:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    @PostMapping
+    public Map<String, String> internationalizeMessages(@RequestBody LocaleUserMessagesRequestDto userMessagesRequest) {
```
@return ResponseBody

patkkka replied:
Regarding this I've found: "Remember, we don’t need to annotate the @RestController-annotated controllers with the @ResponseBody annotation since it’s done by default here."

arturAR commented on frontend/src/main/java/com/java/academy/gym/dto/LocaleUserMessagesRequestDto.java:
```diff
@@ -0,0 +1,28 @@
+    private List<String> messagesKeys;
+
+    public LocaleUserMessagesRequestDto() {
+    }
```
it could be in one line

arturAR commented on frontend/src/main/java/com/java/academy/gym/dto/LocaleUserMessagesRequestDto.java:
```diff
@@ -0,0 +1,28 @@
+public class LocaleUserMessagesRequestDto {
+
+    private String languageCode;
+    private List<String> messagesKeys;
```
It should be map, we want to send this message from client-web to backend and from backend to client-web. 
(in first direction we have empty values)

arturAR commented on frontend/src/main/java/com/java/academy/gym/util/JsonUtil.java:
```diff
@@ -0,0 +1,17 @@
+    private JsonUtil() {
+    }
+
+    public static byte[] toJson(Object object) throws JsonProcessingException {
```
Maybe gson library could be better?

volodymyrd replied:
fasterxml comes with spring so I think you can stay with it

arturAR reviewed (changes requested):
My advices.

mprtcz commented on backend/src/main/java/com/java/academy/gym/model/Locale.java:
```diff
@@ -0,0 +1,76 @@
+    @Size(max = 100)
+    private String country;
+
+    public Locale() {
```
Make it one line

mprtcz commented on backend/src/main/java/com/java/academy/gym/model/UserMessage.java:
```diff
@@ -0,0 +1,73 @@
+
+    private String messageText;
+
+    public UserMessage() {
```
One line here as well

mprtcz commented on frontend/src/main/java/com/java/academy/gym/controller/InternationalizationController.java:
```diff
@@ -0,0 +1,30 @@
+        this.internationalizationService = internationalizationService;
+    }
+
+    @PostMapping
```
Why POST request? I think this one is idempotent.

patkkka replied:
@mprtcz I think you mean that POST is not idempotent.
We had discussion with @arturAR about that and in the first version I implemented this as GET request with params in URL. @arturAR wanted to have POST with params in @RequestBody.
I don't know what approach should be taken here, because in GET request we should't have body (https://stackoverflow.com/questions/978061/http-get-with-request-body etc.). So maybe you can ask @volodymyrd what is correct approach in such situations? For me POST is also weird here...

mprtcz replied:
No, I meant that this method is. And POST is not. So POST should definitely not go here.
[There yo ugo](https://evertpot.com/dropbox-post-api/)
Better to use PATCH or PUT as those methods are idempotent and can have request body. They are not safe, but we will not change anything with this method so it would suffice.
Oh, and apparently [GET can have a body after all](https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol#Summary_table) :suspect:

volodymyrd replied:
GET here is more suitable, but to be honest, we shouldn't spent a lot of time discussing it, it's just theory, I prefer use GET only for read-only operations for the rest one I use POST, because using different methods makes your API more complicated to use and support, and I'm not sure about benefits.

mprtcz commented on frontend/src/test/java/com/java/academy/gym/controller/InternationalizationControllerTest.java:
```diff
@@ -0,0 +1,63 @@
+        Map<String, String> userMessages = new HashMap<>();
+        userMessages.put("MK1", "Message1");
+        userMessages.put("MK2", "Message2");
+        when(mockInternationalizationService.findUserMessagesByKeysAndLanguageCode(any(), any()))
```
I think it's better to specify what messages exactly you'll be passing inside. So instead of any() pass on concrete values.

volodymyrd commented on backend/build.gradle:
```diff
@@ -1,3 +1,9 @@
+ext {
+    versions = [
+            h2: '1.4.196',
```
better move this to the main build.gradle

volodymyrd commented on backend/build.gradle:
```diff
@@ -34,6 +40,7 @@ dependencies {
     compile libs.spring_boot_security
     compile libs.spring_boot_jpa
     compile libs.database_hsqldb
+    compile "com.h2database:h2:${versions.h2}"
```
better move this to the main build.gradle and create libs something like libs.database_h2 and replace with "compile libs.database_h2"

volodymyrd commented on backend/src/main/java/com/java/academy/gym/model/BaseEntity.java:
```diff
@@ -0,0 +1,22 @@
+import javax.persistence.MappedSuperclass;
+
+@MappedSuperclass
+public class BaseEntity {
```
Good idea to create a base entity but better to make it abstract and implement Serializable interface that it helps to use entity as a DTO object

volodymyrd commented on backend/src/main/java/com/java/academy/gym/model/BaseEntity.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    protected Long id;
```
As a best practice is providing @Column annotation with name of field explicitly

volodymyrd commented on backend/src/main/java/com/java/academy/gym/model/BaseEntity.java:
```diff
@@ -0,0 +1,22 @@
+    public void setId(Long id) {
+        this.id = id;
+    }
+}
```
Also I would add here field which responsible for the optimistic lock something like:
@Column(name = "VERSION")
    @Version
    private Integer version;

volodymyrd commented on backend/src/main/java/com/java/academy/gym/model/Locale.java:
```diff
@@ -0,0 +1,76 @@
+
+@Entity
+@Table(name = "locales")
+public class Locale extends BaseEntity {
```
Prefer a natural key to a surrogate key if the natural key is just one column. Here you can create the natural key for example using the "languageCode" field

volodymyrd commented on backend/src/main/java/com/java/academy/gym/model/Locale.java:
```diff
@@ -0,0 +1,76 @@
+
+    @NotNull
+    @Size(max = 10)
+    @Column(unique = true)
```
Instead of or together with "@NotNull" "@Size" and so on annotations better use JPA "@Column" annotation because it affects the database table if you using tools like hbmddl that create DDL scheme. in this case it looks like @Colimn(name = "LANG_CODE", nullable = false, unique = true, length=10)

volodymyrd commented on backend/src/main/java/com/java/academy/gym/service/impl/InternationalizationServiceImpl.java:
```diff
@@ -0,0 +1,37 @@
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(Collectors.toMap(UserMessage::getMessageKey, UserMessage::getMessageText));
+    }
```
It looks like not very optimal, you can calculate how many times it hits database, better to create one query using JOIN

volodymyrd commented on frontend/src/main/java/com/java/academy/gym/dto/LocaleUserMessagesRequestDto.java:
```diff
@@ -0,0 +1,28 @@
+
+import java.util.List;
+
+public class LocaleUserMessagesRequestDto {
```
To be honest I try avoid creating  separated DTO objects and prefer to use the entities, because you keep all things in one place so you get readable, maintainable code without repetition. You can use annotations like @JsonIgnore, @Transient to make it. And don't forget to implement the Serializable interface

patkkka replied:
@volodymyrd In this case we need only language code and messages' keys data in request - so instead of this we should accept whole Locale object and list of UserMessage objects in controller method?

volodymyrd replied:
yes, and you can use the UserMessage as a DTO object you don't need create LocaleUserMessagesRequestDto

volodymyrd commented:
Good start!

patkkka committed:
Fix minor issues from PR comments, move h2 DB to main gradle config

patkkka committed:
Refactor dao, service and controller to handle map with entities as values instead of dto

patkkka commented:
@volodymyrd @mprtcz @arturAR I've implemented nearly all your suggestions, please take a look on that, but still some problems/questions:
1) after usage of @JsonIgnore annotation gradle build fails (you can see logs on travis), I have no idea what is going on, because in IDE compilation is successful and application is working correctly
2) I haven't changed POST method to GET in controller, because I don't know how to test it as Postman doesn't allow to send GET with body (body is grayed out), also when I've read about that, there is a lot of recommendations that GET body (if used) shouldn't have any influence on response and all required data should be included in URL (and that our usage would be incorrect in terms of HTTP specification). So still I don't know how to change it correctly. 
3) Now, when I eliminated DTO and Request body is Map<String, UserMessage>, we will be sending a lot of unnecessary data, which cannot be annotated with @JsonIgnore, because it is needed in response. And it is necessary to send Locale object as component of UserMessage so many times as there is UserMessages in request, not only once. (See JSON examples below)
4) I don't know how to make internationalizeUserMessages method in Service more efficient to eliminate DB query per each user message.
---
Request body example:
```
{
	"MK1": {
		"locale": {
			"languageCode": "pl_PL"
		},
		"messageKey": "MK1",
		"messageText": ""
	},
	"MK2": {
		"locale": {
			"languageCode": "pl_PL"
		},
		"messageKey": "MK2",
		"messageText": ""
	}
}
```

Response body example:
```
{
    "MK1": {
        "locale": {
            "languageCode": "pl_PL",
            "language": "polski",
            "country": "Poland"
        },
        "messageKey": "MK1",
        "messageText": "Message1"
    },
    "MK2": {
        "locale": {
            "languageCode": "pl_PL",
            "language": "polski",
            "country": "Poland"
        },
        "messageKey": "MK2",
        "messageText": "Message2"
    }
}
```

volodymyrd committed:
-fixes build fails after usage of @jsonignore annotation gradle
-adds liquibase support
-adds as example a method retrieving all locals (http://localhost:8080/i18n/locales), creating by liquibase

mprtcz committed:
An attempt to fix failing travis build.

arturAR committed:
Fix Service Tests

mprtcz committed:
Merge pull request #32 from arturAR/fixing-travis

An attempt to fix failing travis build.
